src/libsyntax/parse/parser.rs

```diff
@@ -3480,33 +3480,41 @@ impl<'a> Parser<'a> {
         };
     }
 
-    fn parse_pat_tuple_elements(&mut self, unary_needs_comma: bool)
-                                -> PResult<'a, (Vec<P<Pat>>, Option<usize>)> {
-        let mut fields = vec![];
-        let mut ddpos = None;
+    fn parse_pat_tuple(&mut self) -> PResult<'a, (Vec<P<Pat>>, Option<usize>, bool)> {
```

Comment:
Nit: I'd prefer a name like `parse_pat_parenthesized_list`. Also, a comment explaining this would be nice. Something like:

---

Parses a parenthesized list of patterns like `()`, `(p)`, `(p,)`, `(p, q)`, or `(p, q, ..)`. Returns a vector of the subpatterns that were parsed, along with (a) an option indicating the index of the `..` element and (b) a boolean indicating whether a trailing comma was present. Trailing commas are significant because `(p)` and `(p,)` are different patterns.